### PR TITLE
aws: support EKS Pod Identity

### PR DESCRIPTION
**Pull request by nbaws**
Commit Message: aws: support EKS Pod Identity
Additional Description: Minor patch to allow EKS Pod Identity to be provided via a token file, per the EKS Pod Identity Agent
Risk Level: Low
Testing: Unit
Docs Changes: Yes
Release Notes: Yes
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [inline source/extensions/common/aws/credentials_provider_impl.cc:449]: This is the key part of this change - auth token will take the contents of the token file if path is provided via this environment variable

**Comment by nbaws** [comment]
@suniltheta

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (suniltheta) [inline source/extensions/common/aws/credentials_provider_impl.cc:429]: How about just renaming it to `ContainerCredentialsProvider` based on 
https://docs.aws.amazon.com/sdkref/latest/guide/feature-container-credentials.html ?

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (suniltheta) [inline source/extensions/common/aws/credentials_provider_impl.cc:767]: This can still be const? Since AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE will be provided instead of  AWS_CONTAINER_AUTHORIZATION_TOKEN.

Meanwhile what is the reason not to read `AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE` here? Would that file be dynamically updating?

File: test/extensions/common/aws/credentials_provider_impl_test.cc
Comment (suniltheta) [inline test/extensions/common/aws/credentials_provider_impl_test.cc:2034]: Do you want to set this to `169.254.170.23` as per https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html#pod-id-considerations ?

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [reply source/extensions/common/aws/credentials_provider_impl.cc:767]: The reason behind removing const was because authorization_token provided in ENV variable (for ECS Task Role) is static, however the authorization_token provided in a file for EKS Pod Identity is regularly changing. I will change code path in refresh() instead and change this back to const.

File: test/extensions/common/aws/credentials_provider_impl_test.cc
Comment (nbaws) [reply test/extensions/common/aws/credentials_provider_impl_test.cc:2034]: Good catch :)

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [reply source/extensions/common/aws/credentials_provider_impl.cc:429]: ack

File: test/extensions/common/aws/credentials_provider_impl_test.cc
Comment (nbaws) [inline test/extensions/common/aws/credentials_provider_impl_test.cc:2862]: This appears to be a long standing test bug - we should be clearing these environment variables as part of setup, instead by doing in destructor means that previous environment variable set affects test case 1 (which happened in this new patch)

**Comment by nbaws** [comment]
ty for review @suniltheta

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [inline source/extensions/common/aws/credentials_provider_impl.cc:448]: refactored this piece

**Comment by yanavlasov** [comment]
@nbaws please resolve merge conflict. I will merge after approval from @suniltheta or @lavignes 

/wait

**Review by suniltheta — COMMENTED** [review]
few minor comments

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [inline source/extensions/common/aws/credentials_provider_impl.cc:454]: @suniltheta Per comment on const - authorization_token_ now remains const (ECS use case), however if it is not populated (EKS Pod Identity use case) then it will attempt to read from file instead. If this succeeds, then authorization_header will now contain the file token and will be used. So original const is honoured.

File: test/extensions/common/aws/credentials_provider_impl_test.cc
Comment (nbaws) [reply test/extensions/common/aws/credentials_provider_impl_test.cc:2862]: SetUp is called implicitly for each fixture execution.
Have added comment to indicate.

File: source/extensions/common/aws/credentials_provider_impl.cc
Comment (nbaws) [reply source/extensions/common/aws/credentials_provider_impl.cc:772]: see below for new comment on this.

**Comment by yanavlasov** [comment]
Please merge the main branch after #33115 is submitted to fix CI breakage.

/wait

**Review by suniltheta — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
